desk/app/expose.hoon

```diff
@@ -0,0 +1,2233 @@
+            @
+          ?:  (lth n 10)  "th"
+          $(n (mod n 10))
+        ==
+      --
+    --
+  ::
+  ++  post-from-cite
+    |=  [our=@p now=@da ref=cite:c]
+    ^-  (unit [=nest:g:c =post:d])
+    ?.  ?=(%chan -.ref)
+      ~
+    ::TODO  the whole "deconstruct the ref path" situation is horrendous
+    ?.  ?=([?(%msg %note %curio) @ ~] wer.ref)
+      ~
+    =,  ref
+    =/  base=path
+      %+  weld
+        /(scot %p our)/channels/(scot %da now)
+      /v2/[p.nest]/(scot %p p.q.nest)/[q.q.nest]
+    ?.  .^(? %gu base)  ~
+    :+  ~  nest
+    .^  post:d  %gx
+      %+  weld  base
+      /posts/post/(scot %ud (rash i.t.wer dum:ag))/channel-post-2
+    ==
+  ::
+  ++  update-widget
+    |=  [[our=@p now=@da] open=(set cite:c)]
+    ^-  (list card)
+    ?.  .^(? %gu /(scot %p our)/profile/(scot %da now)/$)
+      ~
+    ~>  %bout.[0 'updating expose widget']
+    =;  widget=[%0 desc=@t %marl marl]
+      =/  =cage  noun+!>([%command %update-widget %groups %expose-all widget])
+      [%pass /profile/widget/all %agent [our %profile] %poke cage]~
+    :^  %0  'Publicized content'  %marl
+    ^-  marl
+    ::
+    =/  cis=(list cite:c)
+      ::REVIEW  maybe limit to the latest n?
+      %+  sort  ~(tap in open)
+      ::  newest first (assumes id nr in path is a timestamp)
+      ::
+      |=  [a=cite:c b=cite:c]
+      ?.  ?=([%chan * ?(%msg %note %curio) @ *] a)  |
+      ?.  ?=([%chan * ?(%msg %note %curio) @ *] b)  &
+      ?~  aa=(rush i.t.wer.a dum:ag)                |
+      ?~  bb=(rush i.t.wer.b dum:ag)                &
+      (gth u.aa u.bb)
+    =/  style=@t
+      '''
+      /* TODO
+         prefer scoping rules to just elements under #groups--expose-all,
+         otherwise style may affect other widgets on the profile
+         (alternatively do inline style attributes)
+      */
+      #groups--expose-all {
+        border: 1px solid red;
+      }
+      #groups--expose-all a {
+        display: block;
+        margin: 1em;
+      }
+      #groups--expose-all .exposed {
+        border: 1px solid black;
+      }
+      '''
+    :-  ;style:"{(trip style)}"
+    %+  murn  cis
+    |=  ref=cite:c
+    ^-  (unit manx)
+    =/  pon=(unit [=nest:g:c =post:d])
+      (post-from-cite our now ref)
+    ?~  pon  ~
+    %-  some
+    =/  link=tape
+      (spud (print:c ref))
+    =,  post.u.pon
+    =/  desc=tape
+      ::TODO  also want to know if there's.. more content, so we can
+      ::      conditionally render the ellipses?
+      (trip (rap 3 (turn (first-inline:u content) flatten-inline:u)))
+    ?-  -.kind-data.post.u.pon
+        %chat
+      ;a.exposed.chat/"/expose{link}"
+        ;span:"{desc}"
+      ==
+    ::
+        %diary
+      ;a.exposed.diary/"/expose{link}"
+        ::TODO  image background?
+        ;h3:"{(trip title.kind-data)}"
+        ;span:"{desc}"
+      ==
+    ::
+        %heap
+      ;a.exposed.heap/"/expose{link}":"TODO gallery item?"
+    ==
+  --
+::
+++  hutils  ::  http request utils
```

Comment:
I pulled these into a lib file in my hm/hooker branch and made profile and hooker point to that, think it's worth it for this too.